### ISSUE TITLE
Incorrect type inference in list comprehension inside closure

### ISSUE THREAD

**Issue by ambroslins** [microsoft/pyright#6067]
**Describe the bug**
Using a variable with a narrowed type loses the narrowing in a list comprehension inside a closure.

**Code**
```python
def f(foo: list | None):
    if isinstance(foo, list):

        def g():
            y: list = foo  # this is fine
            return [x for x in foo]
            #                  ^^^
            # Object of type "None" cannot be used as iterable value
```

**command-line**
pyright 1.1.329 reports this as a type error:
```
> python -m pyright .\test.py
C:\Users\Ambros\Dev\pylance-issue\test.py
  C:\Users\Ambros\Dev\pylance-issue\test.py:6:32 - error: Object of type "None" cannot be used as iterable value (reportOptionalIterable)
1 error, 0 warnings, 0 informations
```

**VS Code extension**
VS Code 1.82.3 with Pylance v2023.9.30 reports the same type error.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
